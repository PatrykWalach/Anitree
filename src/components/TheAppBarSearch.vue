<template>
  <v-col justify="center" class="mx-2">
    <v-form
      v-model="valid"
      :style="{ display: 'contents' }"
      @submit.prevent="submit"
    >
      <v-text-field
        :style="{ 'max-width': '720px' }"
        prepend-inner-icon="search"
        v-model="search"
        autofocus
        clearable
        @click:clear="move"
        @click="move"
        @input="move"
        single-line
        filled
        rounded
        hide-details
        name="search"
        label="Search"
        @keydown.esc="$emit('keydown:esc')"
      ></v-text-field> </v-form
  ></v-col>
</template>
<script lang="ts">
import { ref, Ref, computed, createComponent } from '@vue/composition-api'

export default createComponent({
  setup(_, { emit, root }) {
    const search = ref('')
    const valid = ref(false)

    const rules: Ref<((search: string) => string | true)[]> = ref([
      (search: string) => !!search.length || 'stop it'
    ])

    const submit = async () => {
      if (valid.value && root.$route.query.search !== search.value.trim()) {
        root.$router.push({
          name: 'search',
          query: {
            search: search.value.trim()
          }
        })
        emit('submit')
      }
    }

    const query = computed(() => root.$route.query)

    const isSearched = computed(() => {
      for (const prop in query.value) {
        if (query.value.hasOwnProperty(prop)) {
          return true
        }
      }
      return false
    })

    const move = () => {
      if (root.$route.name !== 'search') {
        root.$router.push({
          name: 'search',
          query: {}
        })
      } else if (isSearched.value) {
        root.$router.replace({
          query: {}
        })
      }
    }

    return { search, valid, rules, submit, move }
  }
})
</script>
