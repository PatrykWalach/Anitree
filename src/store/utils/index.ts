export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
