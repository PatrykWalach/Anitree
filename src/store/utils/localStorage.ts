export const getItem = <T>(key: string, defaultValue?: T) => {
  const cache = localStorage.getItem(key)
  const stored = cache && JSON.parse(cache)

  return {
    ...defaultValue,
    ...stored,
  } as T
}
export const setItem = <T>(key: string, value: T) =>
  localStorage.setItem(key, JSON.stringify({ ...value }))
