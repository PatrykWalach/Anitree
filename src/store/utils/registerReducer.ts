// import * as reducers from '../reducers'
// import { Reducer, ReducersMapObject, combineReducers } from 'redux'
// import { EnhancedStore } from '@reduxjs/toolkit'
// import { store, useStore } from '../'

// const staticReducers: ReducersMapObject<any, any> = reducers
// const asyncReducers: ReducersMapObject<any, any> = {}

// export const registerReducer = <K extends string, R extends Reducer<any>>(
//   key: K,
//   asyncReducer: R,
// ) => {
//   asyncReducers[key] = asyncReducer
//   store.replaceReducer(createReducer(asyncReducers) as any)

//   return () =>
//     useStore<
//       typeof store extends EnhancedStore<infer S>
//         ? EnhancedStore<S & { [key in K]: R }>
//         : any
//     >()
// }

// export const unregisterReducer = <K extends string>(key: K) => {
//   delete asyncReducers[key]
//   store.replaceReducer(createReducer(asyncReducers) as any)
//   return () =>
//     useStore<
//       typeof store extends EnhancedStore<infer S>
//         ? EnhancedStore<Omit<S, K>>
//         : any
//     >()
// }

// const createReducer = (asyncReducers: ReducersMapObject<any, any> = {}) =>
//   combineReducers({
//     ...staticReducers,
//     ...asyncReducers,
//   })
