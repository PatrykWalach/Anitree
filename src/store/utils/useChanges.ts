import {
  DELETE_MEDIA_LIST_ENTRY,
  MEDIA,
  SAVE_MEDIA_LIST_ENTRY,
} from '@/graphql'
import { DeletePending, SavePending } from '../modules/changes'
import {
  DeleteVariables,
  Deleted,
  SaveVariables,
} from '@/graphql/schema/listEntry'
import { DocumentNode, FetchResult } from 'apollo-link'
import { Ref, ref } from '@vue/composition-api'
import { DataProxy } from 'apollo-cache'
import { Media } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import { MutationUpdaterFn } from 'apollo-client'
import { Variables } from '@/graphql/schema/media'
import { useActions } from '..'
import { useDispatch } from 'vue-redux-hooks'
import { useMutation } from '@vue/apollo-composable'

type OnError = (
  fn: (param?: Error | undefined) => void,
) => {
  off: () => void
}

export const handleError = <
  P extends Omit<SavePending, 'id'> | Omit<DeletePending, 'id'>
>(
  onError: OnError,
  type: P['type'],
) => {
  const pending: Ref<P> = ref({ type })

  const dispatch = useDispatch()
  const actions = useActions()

  onError(() => {
    dispatch(actions.changes.UNSHIFT_PENDING(pending.value))
  })

  return pending
}

export const useChange = <D, V>(
  query: DocumentNode,
  update: (cache: DataProxy, { data }: FetchResult<D>, mediaId: number) => void,
) => {
  const { mutate: change, loading, onError, onDone } = useMutation<D, V>(query)
  const mutate = (variables: V, mediaId: number) => {
    change(variables, {
      update: (cache, result) => update(cache, result, mediaId),
    })
  }
  return { loading, mutate, onDone, onError }
}

export const useDelete = () => {
  return useChange<{ DeleteMediaListEntry: Deleted }, DeleteVariables>(
    DELETE_MEDIA_LIST_ENTRY,
    updateDeleteMediaListEntry,
  )
}

export const useSave = () => {
  return useChange<
    {
      SaveMediaListEntry: MediaList
    },
    SaveVariables
  >(SAVE_MEDIA_LIST_ENTRY, updateSaveMediaListEntry)
}

export const useDeleteWithChanges = () => {
  const query = useDelete()

  const pending = handleError(query.onError, 'DELETE')

  const mutate = (variables: DeleteVariables, mediaId: number) => {
    pending.value = {
      mediaId,
      type: 'DELETE',
      variables,
    }

    query.mutate(variables, mediaId)
  }

  return { ...query, mutate }
}
export const useSaveWithChanges = () => {
  const query = useSave()

  const pending = handleError(query.onError, 'SAVE')

  const mutate = (variables: SaveVariables) => {
    pending.value = {
      mediaId: variables.mediaId,
      type: 'SAVE',
      variables,
    }

    query.mutate(variables, variables.mediaId)
  }

  return { ...query, mutate }
}

export const useChanges = () => {
  const { mutate: deleteEntry } = useDelete()
  const { mutate: saveEntry } = useSave()

  const isSavePending = (
    pending: SavePending | DeletePending,
  ): pending is SavePending => pending.type === 'SAVE'

  return { deleteEntry, isSavePending, saveEntry }
}

export const updateSaveMediaListEntry: MutationUpdaterFn<{
  SaveMediaListEntry: MediaList
}> = (cache, { data: result }) => {
  if (!result) return

  const query: {
    variables: Variables
    query: DocumentNode
  } = {
    query: MEDIA,
    variables: { id: result.SaveMediaListEntry.mediaId },
  }

  const data: { Media: Media } | null = cache.readQuery(query)

  if (data) {
    data.Media.mediaListEntry = result.SaveMediaListEntry

    cache.writeQuery({
      ...query,
      data,
    })
  }
}

const updateDeleteMediaListEntry = (
  cache: DataProxy,
  { data: result }: FetchResult<{ DeleteMediaListEntry: Deleted }>,
  id: number,
) => {
  if (!result || !result.DeleteMediaListEntry.deleted) return

  const query: {
    variables: Variables
    query: DocumentNode
  } = {
    query: MEDIA,
    variables: {
      id,
    },
  }

  const data: { Media: Media } | null = cache.readQuery(query)

  if (data) {
    data.Media.mediaListEntry = null

    cache.writeQuery({
      ...query,
      data,
    })
  }
}
