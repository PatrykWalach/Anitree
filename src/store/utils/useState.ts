import { Ref, computed, onUnmounted, ref } from '@vue/composition-api'
import { State } from '..'
import { useStore } from './useStore'

type Option<R> = {
  get: (state: State) => R
  set: (value: R) => void
}

export const useState = <R>(get: Option<R>['get']) => {
  const { subscribe, getState } = useStore()

  const selector: Ref<R> = ref(get(getState()))

  let currentValue: R
  function handleChange() {
    const previousValue = currentValue
    currentValue = get(getState())

    if (previousValue !== currentValue) {
      selector.value = currentValue
    }
  }

  const unsubscribe = subscribe(handleChange)
  onUnmounted(unsubscribe)

  return computed(() => selector.value)
}
