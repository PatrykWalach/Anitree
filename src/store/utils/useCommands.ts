import { ListCommand } from '../modules/commands'
import { computed } from '@vue/composition-api'
import { useState } from './useState'
import { useStore } from './useStore'

export const useCommands = () => {
  const { dispatch, actions } = useStore()

  const history = useState(state => state.commands.history)

  const undo = async () => {
    history.value[history.value.length - 1]
      .undo()
      .then(() => dispatch(actions.commands.POP_COMMAND()))
  }

  const add = async (command: ListCommand) => {
    dispatch(actions.commands.PUSH_COMMAND(command))
    await command.execute()
    return command
  }

  const getPending = computed(() => history.value.filter(({ done }) => !done))
  return { add, getPending, undo }
}
