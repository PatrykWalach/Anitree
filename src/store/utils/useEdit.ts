import { State, useActions } from '../'
import { useDispatch, useSelector } from 'vue-redux-hooks'
import { DollarApollo } from 'vue-apollo/types/vue-apollo'
import { SaveCommand } from '../modules/commands/SaveCommand'
import { SaveVariables } from '@/graphql/schema/listEntry'
import Vue from 'vue'
import { useCommands } from './useCommands'

export const useEdit = () => {
  const dispatch = useDispatch()
  const actions = useActions()
  const { add } = useCommands()

  const close = async () => {
    dispatch(actions.edit.CHANGE_IS_EDITED(false))
    dispatch(actions.edit.RESET_FORM())
  }

  const mediaId = useSelector((state: State) => state.edit.mediaId)
  const form = useSelector((state: State) => state.edit.form)

  const submit = async (apollo: DollarApollo<Vue>) => {
    if (mediaId.value) {
      dispatch(actions.edit.CHANGE_LOADING(true))

      await add(
        new SaveCommand({
          apollo,
          variables: {
            mediaId: mediaId.value,
            ...form.value,
          },
        }),
      )

      dispatch(actions.edit.RESET_FORM())
      dispatch(actions.edit.CHANGE_LOADING(false))
    }
  }

  const changeForm = ({
    form,
    apollo,
  }: {
    form: Partial<SaveVariables>
    apollo: DollarApollo<Vue>
  }) => {
    const syncChanges = useSelector(
      (state: State) => state.settings.syncChanges,
    )
    dispatch(actions.edit.CHANGE_FORM(form))
    if (syncChanges.value) {
      submit(apollo)
    }
  }

  const open = (_mediaId: number) => {
    dispatch(actions.edit.CHANGE_IS_EDITED(true))
    dispatch(actions.edit.CHANGE_MEDIA_ID(_mediaId))
  }

  return {
    changeForm,
    close,
    open,
    submit,
  }
}
