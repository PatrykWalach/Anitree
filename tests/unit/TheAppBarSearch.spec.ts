import TheAppBarSearch from '@/components/TheAppBarSearch.vue'

import { matchSnapshot } from './utils'

describe('TheAppBarSearch', () => {
  matchSnapshot(TheAppBarSearch, {
    mocks: {
      $route: {
        query: {
          search: '',
        },
      },
    },
  })
})
