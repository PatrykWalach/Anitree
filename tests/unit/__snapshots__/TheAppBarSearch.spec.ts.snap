// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheAppBarSearch matches snapshot 1`] = `
<v-container-stub
  pa-0=""
  tag="div"
>
  <v-row-stub
    justify="center"
    nogutters="true"
    tag="div"
  >
    <v-col-stub
      cols="12"
      md="6"
      tag="div"
    >
      <v-form-stub>
        <v-text-field-stub
          backgroundcolor=""
          clearicon="$clear"
          errorcount="1"
          errormessages=""
          filled="true"
          fullwidth="true"
          hidedetails="true"
          label="Search"
          loaderheight="2"
          messages=""
          name="search"
          rounded="true"
          rules=""
          singleline="true"
          successmessages=""
          type="text"
          value=""
        />
      </v-form-stub>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;
